packages/relayer-lib/src/tx_builder/eth_to_cosmos.rs

```diff
@@ -0,0 +1,429 @@
+use ethereum_types::{
+    consensus::sync_committee::compute_sync_committee_period_at_slot,
+    execution::account_proof::AccountProof,
+};
+use ibc_eureka_solidity_types::ics26::router::routerInstance;
+use ibc_proto_eureka::cosmos::tx::v1beta1::TxBody;
+use ibc_proto_eureka::google::protobuf::Any;
+use ibc_proto_eureka::ibc::core::client::v1::{Height, MsgUpdateClient};
+use ibc_proto_eureka::ibc::lightclients::wasm::v1::ClientMessage;
+use ibc_proto_eureka::ibc::{
+    core::{
+        channel::v2::{Channel, QueryChannelRequest, QueryChannelResponse},
+        client::v1::{
+            QueryClientStateRequest, QueryClientStateResponse, QueryConsensusStateRequest,
+            QueryConsensusStateResponse,
+        },
+    },
+    lightclients::wasm::v1::{
+        ClientState as WasmClientState, ConsensusState as WasmConsensusState,
+    },
+};
+use prost::Message;
+use tendermint_rpc::{Client, HttpClient};
+
+use crate::utils::{cosmos, wait_for_condition};
+use crate::{
+    chain::{CosmosSdk, EthEureka},
+    events::EurekaEvent,
+};
+
+use super::r#trait::TxBuilderService;
+
+/// The `TxBuilder` produces txs to [`CosmosSdk`] based on events from [`EthEureka`].
+#[allow(dead_code)]
+pub struct TxBuilder<T: Transport + Clone, P: Provider<T> + Clone> {
+    /// The ETH API client.
+    pub eth_client: EthApiClient<T, P>,
+    /// The Beacon API client.
+    pub beacon_api_client: BeaconApiClient,
+    /// The IBC Eureka router instance.
+    pub ics26_router: routerInstance<T, P>,
+    /// The HTTP client for the Cosmos SDK.
+    pub tm_client: HttpClient,
+    /// The signer address for the Cosmos messages.
+    pub signer_address: String,
+}
+
+impl<T: Transport + Clone, P: Provider<T> + Clone> TxBuilder<T, P> {
+    /// Create a new [`TxBuilder`] instance.
+    pub fn new(
+        ics26_address: Address,
+        provider: P,
+        beacon_api_url: String,
+        tm_client: HttpClient,
+        signer_address: String,
+    ) -> Self {
+        Self {
+            eth_client: EthApiClient::new(provider.clone()),
+            beacon_api_client: BeaconApiClient::new(beacon_api_url),
+            ics26_router: routerInstance::new(ics26_address, provider),
+            tm_client,
+            signer_address,
+        }
+    }
+
+    /// Fetches the eureka channel state from the target chain.
+    /// # Errors
+    /// Returns an error if the channel state cannot be fetched or decoded.
+    pub async fn channel(&self, channel_id: String) -> Result<Channel> {
+        let abci_resp = self
+            .tm_client
+            .abci_query(
+                Some("/ibc.core.channel.v2.Query/Channel".to_string()),
+                QueryChannelRequest { channel_id }.encode_to_vec(),
+                None,
+                false,
+            )
+            .await?;
+
+        QueryChannelResponse::decode(abci_resp.value.as_slice())?
+            .channel
+            .ok_or_else(|| anyhow::anyhow!("No channel state found"))
+    }
```

Comment:
This code is now duplicated in `cosmos_to_cosmos` relayer module as well. So we could refactor this into `TendermintRpcExt`

Reply:
Oh yeah, I also see some client_state I could reuse from there. Also move the generic part of consensus state fetching there for consistency.